Remove `stats._boost` usage by merging it with ``scipy.special._ufuncs`` and using ``scipy.special`` APIs

czgdp1807 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Fixes https://github.com/scipy/scipy/issues/20208
Fixes https://github.com/scipy/scipy/issues/16583

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->

czgdp1807 committed:
MISC: Trim trailing white spaces

czgdp1807 committed:
DEV: Upgrade to latest commit of _boost/math

czgdp1807 committed:
DEV: Use sc.betaincinv for computing stats.beta.ppf and stats.beta.isf

czgdp1807 committed:
DEV: Added survival function and PDF for beta in boost_special_functions.h

czgdp1807 committed:
DEV: Register _beta_sf and _beta_pdf in scipy/special/functions.json

czgdp1807 committed:
DEV: Added docs for _beta_pdf and _beta_sf

czgdp1807 committed:
[skip ci] DEV: Avoid using _boost._beta_* functions

czgdp1807 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -678,33 +678,62 @@ def _pdf(self, x, a, b):
     def _ppf(self, q, a, b):
-        with np.errstate(over='ignore'):  # see gh-17432
-            return _boost._beta_ppf(q, a, b)
+        return sc.betaincinv(a, b, q)
```
So something related to this change. I think PPF of beta distribution is inverse regularised beta incomplete function. However, `_boost._beta_ppf` gives different answers from `sc.betaincinv` for one particular case. Shouldn't they both be giving same behaviour and same answers? I know `_boost._beta_ppf` uses a more general algorithm that is I think it actually takes the inverse of the CDF instead of using the direct result. However, mathematically AFAIK, `sc.betaincinv` should be same as `_boost._beta_ppf`. Am I missing something here?

**`_boost._beta_ppf`**

```zsh
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:49:36) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from scipy import stats
>>> q, a, b = 0.995, 1.0e11, 1.0e13
>>> stats.beta.ppf(q, a, b)
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:699: RuntimeWarning: Error in function boost::math::tgamma<d>(%1%,%1%): Series evaluation exceeded %1% iterations, giving up now.
  return _boost._beta_ppf(q, a, b)
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:699: RuntimeWarning: overflow encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:699: RuntimeWarning: invalid value encountered in _beta_ppf
  return _boost._beta_ppf(q, a, b)
0.009901070347291008
>>> exit()
```

``**sc.betaincinv**``

```zsh
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:49:36) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from scipy import stats
>>> q, a, b = 0.995, 1.0e11, 1.0e13
>>> stats.beta.ppf(q, a, b)
nan
```

dschmitz89 replied:
CC @WarrenWeckesser : do you maybe have an idea about this difference?

czgdp1807 commented:
cc: @rgommers

dschmitz89 reviewed (changes requested):
Thanks for the PR. Boost business is often quite involved, much appreciated!

Does the boost update also fix #20125 and #19348? Acc. to the Boost devs, it should.

For the beta functions: the custom `beta_sf` and `_beta_isf` wrappers can be removed and replaced by [`betaincc`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betaincc.html#scipy.special.betaincc) and [`betainccinv`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betainccinv.html#scipy.special.betainccinv) from special

The beta distribution moments look easy enough judging their analytical formulas. I could not find related issue either, so in my opinion a good idea to replace boost by numpy here.

czgdp1807 commented:
> For the beta functions: the custom `beta_sf` and `_beta_isf` wrappers can be removed and replaced by [`betaincc`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betaincc.html#scipy.special.betaincc) and [`betainccinv`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betainccinv.html#scipy.special.betainccinv) from special

For `isf` I am already using, `sc.betaincinv(a, b, 1 - x)` and it seems to work. Mathematically I think `P(X<=a) = x`  is same as `P(X > a) = 1 - x`, correct?.

For `sf`, I tried `np.subtract(1.0, self._cdf(x, a, b))` (one test fail with this formula) but `scu._beta_sf` gives more accurate results. SciPy docs also say the same that survival function is numerically more accurate than `1.0 - CDF(...)` even though theoretically they both are exactly same. See https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.beta.html (docs of `sf` towards the bottom of the page).

czgdp1807 committed:
DEV: Add invgauss_ppf, invgauss_sf to scipy.special._ufuncs

czgdp1807 committed:
DEV: Use scu._invgauss_isf and scu._invgauss_ppf in invgauss_gen._isf

czgdp1807 committed:
DEV: Use closed form results for ncx2_gen._stats

czgdp1807 commented:
> The beta distribution moments look easy enough judging their analytical formulas. I could not find related issue either, so in my opinion a good idea to replace boost by numpy here.

Yes. They are safe to use and avoids calling into `boost`. 

> Does the boost update also fix #20125 and #19348? Acc. to the Boost devs, it should.

I will have to check on it. Once I replace all the `stats._boost` APIs with `scipy.special` APIs. I will check for these two issues.

czgdp1807 committed:
DEV: Add non-central chi-squared distribution utility functions to scipy.special._ufuncs

dschmitz89 commented:
> > For the beta functions: the custom `beta_sf` and `_beta_isf` wrappers can be removed and replaced by [`betaincc`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betaincc.html#scipy.special.betaincc) and [`betainccinv`](https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.special.betainccinv.html#scipy.special.betainccinv) from special
> 
> For `isf` I am already using, `sc.betaincinv(a, b, 1 - x)` and it seems to work. Mathematically I think `P(X<=a) = x` is same as `P(X > a) = 1 - x`, correct?.
> 
> For `sf`, I tried `np.subtract(1.0, self._cdf(x, a, b))` (one test fail with this formula) but `scu._beta_sf` gives more accurate results. SciPy docs also say the same that survival function is numerically more accurate than `1.0 - CDF(...)` even though theoretically they both are exactly same. See https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.beta.html (docs of `sf` towards the bottom of the page).

The issue with `1 - cdf` is floating point precision. Mathematically, it is of course correct but specialised implementations achieve far better precision. For that reason, there are many `xxxc` functions in `special` (`c` standing for "complementary"=`1-xxx`). See for example the example from the `betaincc` docstring:

![image](https://github.com/scipy/scipy/assets/40656107/69b4cfc1-8fcb-4d9e-aa03-9ca90189caec)

czgdp1807 commented:
Ah! I see. I missed that. Thanks @dschmitz89 :-). I will try using the `sc.betaincc` in `_sf`. If that would work then sure I am +1 on it.

czgdp1807 committed:
DEV: Add non-central F distribution utility functions to scipy.special._ufuncs

czgdp1807 committed:
[skip ci] DEV: Remove scu._beta_sf and replace its usage with sc.betaincc

czgdp1807 commented:
@dschmitz89 So, `sc.betaincc` seems to give work. I am marking your review as addressed. Please feel free to re-open it if there is something I missed.

dschmitz89 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -678,33 +678,62 @@ def _pdf(self, x, a, b):
     def _isf(self, x, a, b):
-        with np.errstate(over='ignore'):  # see gh-17432
-            return _boost._beta_isf(x, a, b)
+        return sc.betaincinv(a, b, 1 - x)
```
```suggestion
        return sc.betainccinv(a, b, x)
```

czgdp1807 committed:
DEV: Add non-central t-distribution utility functions to scipy.special._ufuncs

czgdp1807 committed:
DEV: Replace sc.betaincinv with sc.betainccinv in beta_gen._isf

czgdp1807 committed:
[skip ci] DEV: Add skewnorm distribution utility functions to scipy.special._ufuncs

czgdp1807 commented:
I think we may need to shift stats related wrappers (which I added in this PR) from `boost_special_functions.h` to something like, `boost_stats_functions.h` within the `special` module. Reason being `BOOST_MATH_MAX_ROOT_ITERATION_POLICY`. This root iteration policy is set to `400` in `boost_special_functions.h` which doesn't cause `RuntimeError` for the test `test_ncf_ppf_issue_17026` and hence the test fails. What do you say @dschmitz89 @rgommers ? Everything would be same just `boost_stats_functions.h` getting registered in `functions.json` inside `special`. Let me know.

czgdp1807 committed:
DEV: Add binomial distribution utility functions to scipy.special._ufuncs

DEV: Also added Policy for DRVs and implementations of user_evaluation_error and user_overflow_error

czgdp1807 committed:
DEV: Set options for BOOST

czgdp1807 committed:
DEV: Shift stats related ufuncs to _stats_ufuncs under scipy.special

czgdp1807 committed:
DEV: Shift boost wrappers for stats from scipy.special to scipy.stats

czgdp1807 committed:
[skip ci] DEV: Minimalise wrapper generation code for scipy.stats._ufuncs

czgdp1807 committed:
[skip ci] DEV: Remove stats function docs from scipy.special._add_newdocs.py

czgdp1807 commented:
I am marking this as ready for review. I have created the same pattern for `_ufuncs` generation in `scipy.stats` the way it is being done for `scipy.special` but for statistical distribution. The reason why I couldn't keep those functions in `scipy.special._ufuncs` is because the boost policies for error raising is not the same for `scipy.stats` and `scipy.special`. I tried hacking around it by the but seems like boost policies defined via command line options or by macros doesn't work. The only safe way I could come up with is creation of `scipy.stats._ufuncs` by following the same pattern as `scipy.special._ufuncs`. Keeping stats ufuncs inside special ufuncs didn't feel right to me as well.

All the tests pass except one which need decision on https://github.com/scipy/scipy/pull/20371#discussion_r1549051635

czgdp1807 commented:
I think Github is showing very weird signals. Like I just force pushed my commits (after rebasing on top of `main` branch) and it isn't re-opening my PR. So I created a new one https://github.com/scipy/scipy/pull/20393

rgommers commented:
Yeah, there was an outage earlier today.
